File: correspondence/correspondence/report/meeting_room_schedule/meeting_room_schedule.py
# Copyright (c) 2025, Frappe Technologies and contributors
# For license information, please see license.txt

import frappe
from frappe import _

def execute(filters=None):
	columns = [
		{"fieldname": "booking_id", "label": _("Booking ID"), "fieldtype": "Link", "options": "Meeting Room Booking", "width": 150},
		{"fieldname": "meeting_room", "label": _("Meeting Room"), "fieldtype": "Link", "options": "Meeting Room", "width": 150},
		{"fieldname": "booking_date", "label": _("Date"), "fieldtype": "Date", "width": 100},
		{"fieldname": "start_time", "label": _("Start Time"), "fieldtype": "Time", "width": 100},
		{"fieldname": "end_time", "label": _("End Time"), "fieldtype": "Time", "width": 100},
		{"fieldname": "duration", "label": _("Duration (hrs)"), "fieldtype": "Float", "width": 100},
		{"fieldname": "meeting_title", "label": _("Meeting Title"), "fieldtype": "Data", "width": 200},
		{"fieldname": "booked_by", "label": _("Booked By"), "fieldtype": "Link", "options": "User", "width": 150},
		{"fieldname": "department", "label": _("Department"), "fieldtype": "Link", "options": "Department", "width": 150},
		{"fieldname": "status", "label": _("Status"), "fieldtype": "Data", "width": 100},
		{"fieldname": "number_of_attendees", "label": _("Attendees"), "fieldtype": "Int", "width": 80}
	]
	
	conditions = "WHERE docstatus < 2"
	
	if filters.get("from_date"):
		conditions += f" AND booking_date >= '{filters.get('from_date')}'"
	if filters.get("to_date"):
		conditions += f" AND booking_date <= '{filters.get('to_date')}'"
	if filters.get("meeting_room"):
		conditions += f" AND meeting_room = '{filters.get('meeting_room')}'"
	if filters.get("status"):
		conditions += f" AND status = '{filters.get('status')}'"
	if filters.get("department"):
		conditions += f" AND department = '{filters.get('department')}'"

	data = frappe.db.sql(f"""
		SELECT
			name as booking_id,
			meeting_room,
			booking_date,
			start_time,
			end_time,
			TIMESTAMPDIFF(MINUTE, 
				CONCAT(booking_date, ' ', start_time), 
				CONCAT(booking_date, ' ', end_time)
			) / 60.0 as duration,
			meeting_title,
			booked_by,
			department,
			status,
			number_of_attendees
		FROM
			`tabMeeting Room Booking`
		{conditions}
		ORDER BY
			booking_date DESC, start_time ASC
	""", as_dict=1)

	return columns, data

File: correspondence/correspondence/report/meeting_room_schedule/meeting_room_schedule.js
// Copyright (c) 2025, Frappe Technologies and contributors
// For license information, please see license.txt

frappe.query_reports["Meeting Room Schedule"] = {
    "filters": [
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_days(frappe.datetime.get_today(), -7),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_days(frappe.datetime.get_today(), 30),
            "reqd": 1
        },
        {
            "fieldname": "meeting_room",
            "label": __("Meeting Room"),
            "fieldtype": "Link",
            "options": "Meeting Room"
        },
        {
            "fieldname": "status",
            "label": __("Status"),
            "fieldtype": "Select",
            "options": "\nPending\nApproved\nRejected\nCompleted\nCancelled"
        },
        {
            "fieldname": "department",
            "label": __("Department"),
            "fieldtype": "Link",
            "options": "Department"
        }
    ],

    "formatter": function (value, row, column, data, default_formatter) {
        value = default_formatter(value, row, column, data);

        if (column.fieldname == "status") {
            if (value && value.includes("Approved")) {
                value = `<span class="indicator-pill green">${data.status}</span>`;
            } else if (value && value.includes("Pending")) {
                value = `<span class="indicator-pill orange">${data.status}</span>`;
            } else if (value && value.includes("Rejected")) {
                value = `<span class="indicator-pill red">${data.status}</span>`;
            } else if (value && value.includes("Completed")) {
                value = `<span class="indicator-pill blue">${data.status}</span>`;
            } else if (value && value.includes("Cancelled")) {
                value = `<span class="indicator-pill gray">${data.status}</span>`;
            }
        }

        return value;
    }
};
